### PR TITLE
[489904]  [Formatter2] Partial formatting not working with a single line

### PR DESCRIPTION
**Pull request by cwehrheim**
Fix for bug 489904

If a DSL file is only partially formatted it is possible, that the
ITextReplacement is not fully contained in an allowed ITextRegion but
they overlap.

Modified method 'TextReplacerContext.isInRequestedRange(...)' to not
only check for containment but also for intersection. To check for
intersection I created a private method
'TextReplacerContext.doIntersect(...)'.

Additionally added a test with the example language and formatter
described in the bug.

Signed-off-by: cwehrheim <cwehrheim@itemis.de>

### CODE REVIEW

**Comment by eclipsewebmaster** [comment]
Issue tracker reference:
https://bugs.eclipse.org/bugs/show_bug.cgi?id=489904

**Comment by cwehrheim** [comment]
Hello Moritz. Sorry, I had the wrong email address configured in eclipse. My eclipse account with the signed ECA is registered for christian.wehrheim@itemis.de.

**Comment by cwehrheim** [comment]
Hello Moritz. I removed the custom language, now using the BeeLangTestLanguage.

**Comment by cdietrich** [comment]
ip validation is still broken

**Comment by cwehrheim** [comment]
Hello Moritz, could you please review my commit.

**Comment by cdietrich** [comment]
ping @meysholdt

File: org.eclipse.xtext.testing/src/org/eclipse/xtext/testing/formatter/FormatterTestHelper.java [hunk 1]
```diff
@@ -138,18 +139,45 @@ protected void assertNoSyntaxErrors(XtextResource resource) {
...
-				if (allowed.contains(r))
+		ALLOWED: for (ITextRegion region : regions) {
+			Iterator<ITextReplacement> iterator = rep.iterator();
+			while (iterator.hasNext()) {
+				ITextReplacement replacement = iterator.next();
+				if (region.contains(replacement) || doIntersect(region, replacement))
 					continue ALLOWED;
-				invalid.add(r);
+				
+				if(!iterator.hasNext())
```
Comment (meysholdt) [inline org.eclipse.xtext.testing/src/org/eclipse/xtext/testing/formatter/FormatterTestHelper.java]: this if-statement looks bogus to me. Why should only the last replacement be potentially invalid? The list may not be sorted.

File: org.eclipse.xtext.testing/src/org/eclipse/xtext/testing/formatter/FormatterTestHelper.java [hunk 2]
```diff
@@ -138,18 +139,45 @@ protected void assertNoSyntaxErrors(XtextResource resource) {
...
 			fail("One or more TextReplacements are outside of the allowed region. Region: " + regions, visualized);
 		}
 	}
 
+	private boolean doIntersect(ITextRegion allowed, ITextReplacement replacement) {
+		int allowedOffset = allowed.getOffset();
+		int replacementOffset = replacement.getEndOffset();
+
+		if(allowedOffset == replacementOffset) {
+			return true;
```
Comment (meysholdt) [inline org.eclipse.xtext.testing/src/org/eclipse/xtext/testing/formatter/FormatterTestHelper.java]: *touching* regions don't overlap, do they? I think this case shouldn't be considered true.

File: org.eclipse.xtext.tests/src/org/eclipse/xtext/formatting2/partialformatting/PartialFormattingTest.xtend [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+		unit unitname {
+				source : /a/b/c/;
+			output : /d/e/f/;
+		}
+		'''.toString
+		val part = "		source : /a/b/c/;"
+		val offset = source.indexOf(part)
+		val length = part.length
+		
+		// does not work
```
Comment (meysholdt) [inline org.eclipse.xtext.tests/src/org/eclipse/xtext/formatting2/partialformatting/PartialFormattingTest.xtend]: If this test isn't green, please don't submit it for review.

File: org.eclipse.xtext.tests/src/org/eclipse/xtext/formatting2/partialformatting/PartialFormattingTest.xtend [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+/**
+ * @author cwehrheim - Initial contribution and API
+ */
+@RunWith(XtextRunner)
+@InjectWith(MyPartialFormattingTestLanguageInjectorProvider)
+class PartialFormattingTest {
+	@Inject extension GenericFormatterTester
+
+	@Test
+	def void testFormatting1() {
```
Comment (meysholdt) [inline org.eclipse.xtext.tests/src/org/eclipse/xtext/formatting2/partialformatting/PartialFormattingTest.xtend]: This is the test that is already working without your change, isn't it? I think we don't need this one.

Instead, there should be a test for the second scenario you're implementing: TextReplacements overlapping `allowedRegion.endOffset`.

File: org.eclipse.xtext/src/org/eclipse/xtext/formatting2/internal/TextReplacerContext.java [hunk 5]
```diff
@@ -181,12 +181,34 @@ public boolean isAutowrap() {
...
-					&& region.getOffset() + region.getLength() >= repl.getEndOffset())
+		for(ITextRegion allowed : regions) {
+			if(allowed.contains(repl) || doIntersect(allowed, repl)) {
 				return true;
+			}
+		}
+		return false;
+	}
+	
+	private boolean doIntersect(ITextRegion allowed, ITextReplacement replacement) {
```
Comment (meysholdt) [inline org.eclipse.xtext/src/org/eclipse/xtext/formatting2/internal/TextReplacerContext.java]: Can this redundant code be avoided?

**Comment by szarnekow** [comment]
@meysholdt Could you please take another look?

**Comment by cdietrich** [comment]
http://services.typefox.io/open-source/jenkins/job/xtext-web/job/bug489904_partial_formatting/2/

**Pull request closed without merging**